crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -382,14 +389,97 @@ impl CheckpointExecutor {
         let _parallel_step_guard =
             mysten_metrics::monitored_scope("CheckpointExecutor::parallel_step");
 
-        let (mut ckpt_state, unexecuted_tx_digests) = tokio::task::spawn_blocking({
+        // Note: only the fullnode path has end-of-epoch logic.
+        if self.state.is_fullnode(&self.epoch_store) || checkpoint.is_last_checkpoint_of_epoch() {
+            self.execute_checkpoint_fullnode(checkpoint).await
```

Comment:
IIUC correctly the "fullnode" path will still run sometimes on validators? in that case can we give these more accurate names?

strawman ideas, you can likely come up with better:
execute_checkpoint_incremental
execute_checkpoint_fast